### PR TITLE
Social | Add disconnect/reconnect UI and logic

### PR DESCRIPTION
**Pull request by manzoorwanijk**
## Proposed changes:

* Wire up Disconnect button
* Add Reconnect button for broken connections

### Other information:

- [ ] Have you written new tests for your changes, if applicable?
- [ ] Have you checked the E2E test CI results, and verified that your changes do not break them?
- [ ] Have you tested your changes on WordPress.com, if applicable (if so, you'll see a generated comment below with a script to run)?

## Testing instructions:

You can switch between the new and old UI with `define( 'JETPACK_SOCIAL_USE_ADMIN_UI_V1', true );
` in your wp-config file

  * MERGED - Ensure that your site points to your sandbox with D146900-code
  * Go to the social admin page
  *  Confirm that you see the Disconnect buttons for the connections
  * Click on Disconnect
  * Confirm that the connection is disconnected
  * Confirm that you see the notice for the same
  * Reload the page
  * Confirm that the disconnected connection is not in the list

https://github.com/Automattic/jetpack/assets/18226415/b1a8a1be-afbd-4e03-9964-a9f2f32eeccd

Screenshot 2024-05-07 at 3 02 19 PM
Screenshot 2024-05-07 at 3 03 35 PM
Screenshot 2024-05-07 at 3 04 15 PM

### CODE REVIEW

File: projects/packages/publicize/src/class-publicize.php [hunk 1]
```diff
@@ -259,7 +259,7 @@ public function get_all_connections_for_user() {
 								array(
 									'service_name'   => $service_name,
 									'connection_id'  => $connection['connection_data']['id'],
-									'can_disconnect' => current_user_can( 'manage_options' ) || get_current_user_id() === $user_id,
+									'can_disconnect' => current_user_can( 'edit_others_posts' ) || get_current_user_id() === $user_id,
```
Comment (manzoorwanijk) [inline projects/packages/publicize/src/class-publicize.php:262]: This is the permission that [Calypso uses](https://github.com/Automattic/wp-calypso/blob/c1cb83c9aa061c8f7b197c97bbe19d1c2ad0c19c/client/state/sharing/publicize/selectors.js#L87).

File: projects/js-packages/publicize-components/src/social-store/types.ts [hunk 2]
```diff
@@ -7,6 +7,8 @@ export type SharesData = {
 	shared_posts_count: number;
 };
 
+export type ConnectionStatus = 'ok' | 'must_reauth' | 'invalid' | 'broken' | 'refresh-failed';
```
Comment (manzoorwanijk) [inline projects/js-packages/publicize-components/src/social-store/types.ts:10]: Those are the statuses I found in `external-connections.php` - fbhepr%2Skers%2Sjcpbz%2Sjc%2Qpbagrag%2Syvo%2Srkgreany%2Qpbaarpgvbaf.cuc%3Se%3Q97ooo5qs%231203-og

**Comment by manzoorwanijk** [comment]
Why is @Automattic/jetpack-reach team not shown in the Request for review?

File: projects/js-packages/publicize-components/src/components/connection-management/disconnect.tsx [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+			size="small"
+			onClick={ onClickDisconnect }
+			disabled={ isDisconnectingAny }
+			variant={ variant }
+			isDestructive={ isDestructive }
+		>
+			{ isDisconnectingThis
+				? __( 'Disconnecting…', 'jetpack' )
+				: label || _x( 'Disconnect', 'Disconnect a social media account', 'jetpack' ) }
+		</Button>
```
Comment (gmjuhasz) [inline projects/js-packages/publicize-components/src/components/connection-management/disconnect.tsx]: I like the idea of showing them that they are disconnecting at the moment, but I dislike that the table is jumping width during it. Maybe we could increase the size of the cell so it still has space to expand, without expanding the column, or just leave the button title as `Disconnect` and disable it

CleanShot 2024-05-07 at 20 52 30 png

File: projects/js-packages/publicize-components/src/components/connection-management/snackbars.tsx
Comment (gmjuhasz) [inline projects/js-packages/publicize-components/src/components/connection-management/snackbars.tsx:1]: I love this component, but I'm also a bit worried about making the UI jump a lot with all the notifications. Could we overlay the notification somewhere, or use general notifications similar to what we do in Calypso you think? 🤔 

CleanShot 2024-05-07 at 20 55 14 png

File: projects/js-packages/publicize-components/src/components/connection-management/connection-status.tsx [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+
+/**
+ * Connection status component
+ *
+ * @param {ConnectionStatusProps} props - component props
+ *
+ * @returns {import('react').ReactNode} - React element
+ */
+export function ConnectionStatus( { connection, onReconnect }: ConnectionStatusProps ) {
+	if ( connection.status === 'ok' ) {
```
Comment (gmjuhasz) [inline projects/js-packages/publicize-components/src/components/connection-management/connection-status.tsx]: This is never happens to be the case for me and all the connections show as broken

File: projects/js-packages/publicize-components/src/components/connection-management/snackbars.tsx [hunk 5]
```diff
@@ -0,0 +1,28 @@
+import { SnackbarList } from '@wordpress/components';
+import { useSelect, useDispatch } from '@wordpress/data';
+import { store as noticesStore } from '@wordpress/notices';
+import styles from './style.module.scss';
+
+// Last three notices. Slices from the tail end of the list.
+const MAX_VISIBLE_NOTICES = -3;
```
Comment (gmjuhasz) [inline projects/js-packages/publicize-components/src/components/connection-management/snackbars.tsx]: Nitpick, but maybe for readability of the variable name, we could have this as `3` and negate it in the slicing itself?

**Review by gmjuhasz — COMMENTED** [review]
This looks good, I just had a couple issues while testing:

- While disconnecting, the UI says that the response is not a valid json, the request returns nothing, however the disconnection works because on refresh the connection gets removed.

- For some reason, all of my connections are flagged as broken and I see the reconnect button. 🤔  I double checked, and the site is sandboxed with Sid's change.

- I get an issue on the Jetpack settings page:
CleanShot 2024-05-07 at 21 43 43 png

Maybe we can pair up on this tomorrow?

File: projects/js-packages/publicize-components/src/components/connection-management/disconnect.tsx (hunk 3 above)
Comment (gmjuhasz) [reply projects/js-packages/publicize-components/src/components/connection-management/disconnect.tsx]: Or, another idea, to show a spinner in the disabled button while it's doing it's thing?

Comment (manzoorwanijk) [reply projects/js-packages/publicize-components/src/components/connection-management/disconnect.tsx]: Yeah, I didn't spend much time polishing the UI. It definitely is not the best. Needs many improvements as you pointed out.

File: projects/js-packages/publicize-components/src/components/connection-management/snackbars.tsx
Comment (manzoorwanijk) [reply projects/js-packages/publicize-components/src/components/connection-management/snackbars.tsx:1]: Yeah, we can do that but I wanted something that we could use anywhere in future. For example we want to use this same UI in editor. So, having inline notifications will work better for that.

File: projects/js-packages/publicize-components/src/components/connection-management/connection-status.tsx (hunk 4 above)
Comment (manzoorwanijk) [reply projects/js-packages/publicize-components/src/components/connection-management/connection-status.tsx]: Yeah, the reason for that is the connection data is incomplete as of now unless this is done https://github.com/Automattic/jetpack-reach/issues/310

File: projects/js-packages/publicize-components/src/components/connection-management/snackbars.tsx (hunk 5 above)
Comment (manzoorwanijk) [reply projects/js-packages/publicize-components/src/components/connection-management/snackbars.tsx]: Yeah, sure. I actually copied this component as is from gutenberg.

**Comment by manzoorwanijk** [comment]
> * While disconnecting, the UI says that the response is not a valid json, the request returns nothing, however the disconnection works because on refresh the connection gets removed.

Yes, because the WPCOM endpoint doesn't return anything on connection deletion. Sid is working on that I think

> * For some reason, all of my connections are flagged as broken and I see the reconnect button. 🤔  I double checked, and the site is sandboxed with Sid's change.

Yes, because we don't get the `status` field for connections yet. We need this https://github.com/Automattic/jetpack-reach/issues/310

> * I get an issue on the Jetpack settings page:

Did you try to run build?

> Maybe we can pair up on this tomorrow?

Sure

**Comment by manzoorwanijk** [comment]
Here is the updated UI


https://github.com/Automattic/jetpack/assets/18226415/cd5e96bc-07f7-4213-9adf-bd8e064e322e

**Review by gmjuhasz — APPROVED** [review]
This is great!

**Pull request merged**